compiler/src/dotty/tools/dotc/transform/SpecializeFunctions.scala

```diff
@@ -0,0 +1,158 @@
+      def newParents = tp.parents.mapConserve { parent =>
+        if (defn.isPlainFunctionClass(parent.symbol)) {
+          val typeParams = tp.typeRef.baseArgTypes(parent.classSymbol)
+          val interface = specInterface(typeParams)
+
+          if (interface.exists) {
+            val specializedApply: Symbol = {
+              val specializedMethodName = specializedName(nme.apply, typeParams)
+              ctx.newSymbol(
+                sym,
+                specializedMethodName,
+                Flags.Override | Flags.Method,
+                interface.info.decls.lookup(specializedMethodName).info
+              )
+            }
+
+            newDecls.enter(specializedApply)
+            interface.typeRef
+          }
+          else parent
+        }
+        else parent
+      }
+
+      tp.derivedClassInfo(classParents = newParents, decls = newDecls)
+    }
+
+    case _ => tp
+  }
+
+  override def transformTemplate(tree: Template)(implicit ctx: Context, info: TransformerInfo) = {
+    val buf = new mutable.ListBuffer[Tree]
+    val newBody = tree.body.mapConserve {
+      case dt: DefDef if dt.name == nme.apply && dt.vparamss.length == 1 => {
+        val specializedApply = ctx.owner.info.decls.lookup {
+          specializedName(
+            nme.apply,
+            dt.vparamss.head.map(_.symbol.info) :+ dt.tpe.widen.finalResultType
+          )
+        }
+
+        if (specializedApply.exists) {
+          val apply = specializedApply.asTerm
+          val specializedDecl =
+            polyDefDef(apply, trefs => vrefss => {
+              dt.rhs
+                .changeOwner(dt.symbol, apply)
+                .subst(dt.vparamss.flatten.map(_.symbol), vrefss.flatten.map(_.symbol))
+            })
+
+          buf += specializedDecl
+
+          // create a forwarding to the specialized apply
+          cpy.DefDef(dt)(rhs = {
+            tpd
+            .ref(apply)
+            .appliedToArgs(dt.vparamss.head.map(vparam => ref(vparam.symbol)))
+          })
+        } else dt
+      }
+      case x => x
+    }
+
+    val newParents = tree.parents.mapConserve { parent =>
+        if (defn.isPlainFunctionClass(parent.symbol)) {
+          val typeParams = tree.tpe.baseArgTypes(parent.symbol)
+          val interface = specInterface(typeParams)
+
+          if (interface.exists) TypeTree(interface.info)
+          else parent
+        }
+        else parent
+    }
+
+    cpy.Template(tree)(parents = newParents, body = buf.toList ++ newBody)
+  }
+
+  /** Dispatch to specialized `apply`s in user code */
+  override def transformApply(tree: Apply)(implicit ctx: Context, info: TransformerInfo) = {
+    import ast.Trees._
+    tree match {
+      case Apply(select @ Select(id, nme.apply), arg :: Nil) => {
+        val params = List(arg.tpe, tree.tpe)
+        val specializedApply = specializedName(nme.apply, params)
+        val hasOverridenSpecializedApply = id.tpe.decls.iterator.exists {
```

Comment:
why iterate through all by hand instead of performing member lookup?
Why do you need to check if `sym is override`?